AX: Tree updates become broken when children change for a dynamically ignored element and its unignored ancestor in the same tree update cycle

twilco opened the pull request:
#### c630047b889cdac6aaabf58285f7454af91bb7e8
<pre>
AX: Tree updates become broken when children change for a dynamically ignored element and its unignored ancestor in the same tree update cycle
<a href="https://bugs.webkit.org/show_bug.cgi?id=281324">https://bugs.webkit.org/show_bug.cgi?id=281324</a>
<a href="https://rdar.apple.com/137765781">rdar://137765781</a>

Reviewed by Chris Fleizach.

Consider this HTML:

```
&lt;main&gt;
++&lt;div tabindex=&quot;0&quot;&gt;Foo&lt;/div&gt;
&lt;/main
```

And this sequence:

  1. Insert one new child to the main and the div

  2. Remove tabindex from the div (causing `AXIsolatedTree::updateNode` for the div, in turn queuing the div in
     `AXIsolatedTree::m_unresolvedPendingAppends`, thus protecting it from being removed from `AXIsolatedTree::m_nodeMap`
     in this tree update cycle). This also means the div will go from being unignored to ignored in the middle of handling
     the children change for it (`AXObjectCache::handleChildrenChanged`)

  3. In `AXIsolatedTree::updateChildren` for the main, we correctly realize that we no longer have the div as a child
     because it is now ignored, and thus try to remove it from the node map. But we protected it in step 2, meaning
     it won&apos;t be removed, despite it not &quot;being in the tree&quot; (which is what the node map is supposed to model).

  4. In any subsequent insert into the subtree of the div, we will fail to add said content to the accessibility tree.
     This is because in `AXIsolatedTree::updateChildren`, we use the `m_nodeMap` to iterate up to the nearest &quot;in-tree&quot;
     (unignored) ancestor, and update that. But because we never removed the div from the node map, it will always receive
     the updates, not our main.

We fix this by immediately removing anything becomes ignored from the node map, guaranteeing it can&apos;t have improper
effect on children updates that may be in-progress.

Two layout tests are added. These tests only differ in the is-ignored conditions they exercise, ensuring that if we
change or remove one, we don&apos;t lose test coverage of this bug.

This patch also includes an unrelated change to add the ability to log AXPropertyMap and AXPropertyName via WTF::TextStream,
as I found this useful when investigating this bug.

* LayoutTests/accessibility/tree-update-with-dynamically-ignored-inline-child-div-expected.txt: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-inline-child-div.html: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-tabindex-div-expected.txt: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-tabindex-div.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new tests.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectBecameIgnored):

Canonical link: <a href="https://commits.webkit.org/285073@main">https://commits.webkit.org/285073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d9e2639ae48e0642e3c500e0a11c5da2244376c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24218 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75556 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22650 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22469 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56425 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14898 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74511 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46156 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61540 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36867 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19005 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20991 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64718 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19371 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77275 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15679 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64139 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15722 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61580 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64132 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15801 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12283 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5917 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46658 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1437 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47729 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49013 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -543,6 +543,14 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     Ref<Document> protectedDocument() const;
     constexpr const std::optional<PageIdentifier>& pageID() const { return m_pageID; }
 
+#if !ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE) && ENABLE(ACCESSIBILITY_ISOLATED_TREE)
```
Should we put a `FIXME` in front of one of these that `objectBecameIgnored` can be removed once `INCLUDE_IGNORED_IN_CORE_AX_TREE` is enabled?

twilco replied:
IMO that's implicit, as when `INCLUDE_IGNORED_IN_CORE_AX_TREE` is removed and defaulted to be "on", any code that is `!ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)` only has no reason to exist 🤔

hoffmanjoshua replied:
Fair point, we can leave it out

twilco committed:
AX: Tree updates become broken when children change for a dynamically ignored element and its unignored ancestor in the same tree update cycle
https://bugs.webkit.org/show_bug.cgi?id=281324
rdar://137765781

Reviewed by Chris Fleizach.

Consider this HTML:

```
<main>
++<div tabindex="0">Foo</div>
</main
```

And this sequence:

  1. Insert one new child to the main and the div

  2. Remove tabindex from the div (causing `AXIsolatedTree::updateNode` for the div, in turn queuing the div in
     `AXIsolatedTree::m_unresolvedPendingAppends`, thus protecting it from being removed from `AXIsolatedTree::m_nodeMap`
     in this tree update cycle). This also means the div will go from being unignored to ignored in the middle of handling
     the children change for it (`AXObjectCache::handleChildrenChanged`)

  3. In `AXIsolatedTree::updateChildren` for the main, we correctly realize that we no longer have the div as a child
     because it is now ignored, and thus try to remove it from the node map. But we protected it in step 2, meaning
     it won't be removed, despite it not "being in the tree" (which is what the node map is supposed to model).

  4. In any subsequent insert into the subtree of the div, we will fail to add said content to the accessibility tree.
     This is because in `AXIsolatedTree::updateChildren`, we use the `m_nodeMap` to iterate up to the nearest "in-tree"
     (unignored) ancestor, and update that. But because we never removed the div from the node map, it will always receive
     the updates, not our main.

We fix this by immediately removing anything becomes ignored from the node map, guaranteeing it can't have improper
effect on children updates that may be in-progress.

Two layout tests are added. These tests only differ in the is-ignored conditions they exercise, ensuring that if we
change or remove one, we don't lose test coverage of this bug.

This patch also includes an unrelated change to add the ability to log AXPropertyMap and AXPropertyName via WTF::TextStream,
as I found this useful when investigating this bug.

* LayoutTests/accessibility/tree-update-with-dynamically-ignored-inline-child-div-expected.txt: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-inline-child-div.html: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-tabindex-div-expected.txt: Added.
* LayoutTests/accessibility/tree-update-with-dynamically-ignored-tabindex-div.html: Added.
* LayoutTests/platform/glib/TestExpectations: Skip new tests.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::objectBecameIgnored):

Canonical link: https://commits.webkit.org/285073@main
